blueprints/flexible-topologies.md

```diff
@@ -0,0 +1,395 @@
+            "parents": []
+        }
+    ]
+}
+```
+
+response JSON:
+```JSON
+{
+    "alerts": [
+        {
+            "text": "topology was created successfully",
+            "level": "success"
+        }
+    ],
+    "response": {
+        "name": "foo",
+        "description": "a foo topology",
+        "nodes": [
+            {
+                "cachegroup": "child-cachegroup",
+                "parents": [1, 2]
+            },
+            {
+                "cachegroup": "parent-cachegroup",
+                "parents": []
+            },
+            {
+                "cachegroup": "secondary-parent-cachegroup",
+                "parents": []
+            }
+        ]
+    }
+}
+
+```
+
+##### `PUT /topologies?name=foo`
+
+request JSON:
+```JSON
+{
+    "name": "foo",
+    "description": "a foo topology",
+    "nodes": [
+        {
+            "cachegroup": "child-cachegroup",
+            "parents": [1, 2]
+        },
+        {
+            "cachegroup": "parent-cachegroup",
+            "parents": []
+        },
+        {
+            "cachegroup": "secondary-parent-cachegroup",
+            "parents": []
+        }
+    ]
+}
+```
+
+response JSON:
+```JSON
+{
+    "alerts": [
+        {
+            "text": "topology was updated successfully",
+            "level": "success"
+        }
+    ],
+    "response": {
+        "name": "foo",
+        "description": "a foo topology",
+        "nodes": [
+            {
+                "cachegroup": "child-cachegroup",
+                "parents": [1, 2]
+            },
+            {
+                "cachegroup": "parent-cachegroup",
+                "parents": []
+            },
+            {
+                "cachegroup": "secondary-parent-cachegroup",
+                "parents": []
+            }
+        ]
+    }
+}
+```
+
+##### `DELETE /topologies?name=foo`
+
+response JSON:
+```JSON
+{
+    "alerts": [
+        {
+            "text": "topology was deleted successfully",
+            "level": "success"
+        }
+    ]
+}
+```
+
+##### `/deliveryservices` endpoints
+
+All relevant Delivery Service APIs will have their JSON request and response objects modified to include a new `topology` field which references the name of the topology it's assigned to:
+```JSON
+{
+    ...
+    "topology": "foo"
+}
+```
+
+##### The various `/snapshot` endpoints
+
+The various `/snapshot` endpoints will need to be updated to include new Topologies data along with their associations to Delivery Services in the `CRConfig.json` snapshot. The data should only include the `EDGE_LOC` cachegroups of the Topologies, because those are all Traffic Router needs.
+
+##### Various endpoints that are affected by cachegroup parentage or deliveryservice-server assignment
+
+API endpoints that do things such as the following may need to be updated to take Topology-based Delivery Service assignment and parentage into account:
+- assign a Delivery Service to a server (or vice versa)
```

Comment:
I would suggest leaving all the ds/server and server/ds endpoints as is. We can circle back and mark them as deprecated at some point. Technically, they are still relevant as you can either a) assign a topology to a ds or b) assign servers to a ds.

In my opinion, there could be 2 ways to fetch servers "assigned to a ds"

1. GET /ds/{id}/servers (existing) which will return servers DIRECTLY assigned to a ds
2. GET /servers?topology=foo which will return servers INDIRECTLY assigned to a ds via a topology.

We can make TP smart enough to use the correct one based on if a ds has a topology or not and we can update documentation to be clear what each endpoint is for.

Reply:
Yes but a Delivery Service cannot simultaneously have an assigned Topology and assigned Servers. So the Delivery Service-to-server assignment endpoints will need to check if a Delivery Service has a Topology before allowing servers to be assigned.

Reply:
so what's the topology equivalent of this:

GET servers/{id}/deliveryservices?

hmm. maybe 

GET /topologies?server=id would give you all the topologies that include the server...but how to fetch all the delivery services for those topologies...?

i guess my ultimate point is leave the ds/server server/ds endpoints as is and come up with other topology related ones to achieve the same result from a GET perspective.

Reply:
> Yes but a Delivery Service cannot simultaneously have an assigned Topology and assigned Servers. So the Delivery Service-to-server assignment endpoints will need to check if a Delivery Service has a Topology before allowing servers to be assigned.

actually, i think it can have both....but if you have both, topology wins.